model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java

```diff
@@ -0,0 +1,344 @@
+    }
+
+    private MapRoleEntity registerEntityForChanges(MapRoleEntity origEntity) {
+        final MapRoleEntity res = Serialization.from(origEntity);
+        tx.putIfChanged(origEntity.getId(), res, MapRoleEntity::isUpdated);
+        return res;
+    }
+
+    private Predicate<MapRoleEntity> entityRealmFilter(RealmModel realm) {
+        if (realm == null || realm.getId() == null) {
+            return MapRoleProvider.ALWAYS_FALSE;
+        }
+        String realmId = realm.getId();
+        return entity -> Objects.equals(realmId, entity.getRealmId());
+    }
+
+    private Predicate<MapRoleEntity> entityClientFilter(ClientModel client) {
+        if (client == null || client.getId() == null) {
+            return MapRoleProvider.ALWAYS_FALSE;
+        }
+        String clientId = client.getId();
+        return entity -> entity.isClientRole() && 
+                Objects.equals(clientId, entity.getClientId());
+    }
+
+    private Stream<MapRoleEntity> getNotRemovedUpdatedRolesStream() {
+        Stream<MapRoleEntity> updatedAndNotRemovedRolesStream = roleStore.entrySet().stream()
+          .map(tx::getUpdated)    // If the role has been removed, tx.get will return null, otherwise it will return me.getValue()
+          .filter(Objects::nonNull);
+        return Stream.concat(tx.createdValuesStream(roleStore.keySet()), updatedAndNotRemovedRolesStream);
+    }
+
+    @Override
+    public RoleModel addRealmRole(RealmModel realm, String id, String name) {
+        final UUID entityId = id == null ? UUID.randomUUID() : UUID.fromString(id);
+
+        LOG.tracef("addRealmRole(%s, %s, %s)%s", realm.getName(), id, name, getShortStackTrace());
+
+        MapRoleEntity entity = new MapRoleEntity(entityId, realm.getId());
+        entity.setName(name);
+        entity.setRealmId(realm.getId());
+        if (tx.get(entity.getId(), roleStore::get) != null) {
+            throw new ModelDuplicateException("Role exists: " + id);
+        }
+        tx.putIfAbsent(entity.getId(), entity);
+        return entityToAdapterFunc(realm).apply(entity);
+    }
+
+    @Override
+    public Stream<RoleModel> getRealmRolesStream(RealmModel realm, Integer first, Integer max) {
+        Stream<RoleModel> s = getRealmRolesStream(realm);
+        if (first != null && first >= 0) {
+            s = s.skip(first);
+        }
+        if (max != null && max >= 0) {
+            s = s.limit(max);
+        }
+        return s;
+    }
+
+    @Override
+    public Stream<RoleModel> getRealmRolesStream(RealmModel realm) {
+        return getNotRemovedUpdatedRolesStream()
+                .filter(entityRealmFilter(realm))
+                .filter(this::isRealmRole)
+                .sorted(COMPARE_BY_NAME)
+                .map(entityToAdapterFunc(realm));
+    }
+
+    private boolean isRealmRole(MapRoleEntity role) {
+        return ! role.isClientRole();
+    }
+
+    @Override
+    public RoleModel addClientRole(ClientModel client, String id, String name) {
```

Comment:
Here we should check that there is no client role with the name for the client. This is enforced [here](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/entities/RoleEntity.java#L55) for the jpa layer.